crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -201,13 +201,89 @@ impl SettlementEncoder {
                 )?
             }
             OrderClass::Limit => {
-                anyhow::bail!("limit orders are not supported by the SettlementEncoder");
+                // Solvers calculate with slightly adjusted amounts compared to this order but because
+                // limit orders are fill-or-kill we can simply use the total original `sell_amount`.
+                let executed_amount = match order.data.kind {
+                    OrderKind::Sell => order.data.sell_amount,
+                    OrderKind::Buy => order.data.buy_amount,
+                };
+                let buy_price = self.custom_price_for_limit_order(&order)?;
+
+                self.add_custom_price_trade(
+                    order,
+                    executed_amount,
+                    scaled_unsubsidized_fee,
+                    buy_price,
+                )?
             }
         };
         self.pre_interactions.extend(interactions.pre.into_iter());
         Ok(execution)
     }
 
+    /// Uses the uniform clearing prices to compute the individual buy token price to satisfy the
+    /// original limit order which was adjusted to account for the `surplus_fee` (see
+    /// `compute_synthetic_order_amounts_if_limit_order()`).
+    /// Returns an error if the UCP doesn't contain the traded tokens or if under- or overflows
+    /// happen during the computation.
+    fn custom_price_for_limit_order(&self, order: &Order) -> Result<U256> {
```

Comment:
Maybe to make this easier to test we can pass in the uniform prices you get from `self` so that we have a freestanding function. Then the standalone tests also makes the behavior easier to understand and can be combined with `compute_synthetic_order_amounts_if_limit_order`.

Reply:
I recognize it's something we disagree on.